xhtml2pdf/w3c/cssSpecial.py

```diff
@@ -242,7 +242,7 @@ def parseSpecialRules(declarations, debug=0):
             part = getNextPart(parts) or oparts
             if part:
 
-                if hasattr(part, '__iter__') and (type("." in part) or ("data:" in part)):
+                if hasattr(part, '__iter__') and (("." in part) or ("data:" in part)):
                     dd.append(("background-image", part, last))
```

Comment:
A small fix to properly detect the "background-image" property. The call to type() is wrong here. I realized this while testing the background-image property.

Reply:
@pedroszg maybe you can help me out with this one:

The unit tests are passing my pull request. But the testrender is failing. This is the only line in the code base that I've change. The rest are changes to unit tests.

After a quick check in the Travis log, it seems like this file is the problem: "test-loremipsum.html".

We have some table headers with a background color in the corresponding css\content.css file. When I check the reference files "test-loremipsum.page1.png" and "test-loremipsum.page2.png" then the background-color is missing in the headers. So it seems like the reference files are actually wrong here.  But this is just a guess. Can you maybe run the testrender with this small change and then let me know what the differences are?

I still couldn't get imagemagick running for the testrender, so I can't check if myself. Maybe you can help me out here. Thanks! :)

Reply:
Hi @fbernhart  we make a PR #525  with your suggestion about the Type() to complete it, also we updated the referecence dir with the new images.
Can you explain me with more details your issue with imagemagick?

Reply:
@pedroszg  Thanks a lot for your PR! 

I got imagemagick running now. I had to point testrender.py to the right directory, where imagemagick is.

But still I couldn't get the testrender fully running. I was able to create a local reference. But then, when I want to compare the images, I get the following:

Rendering E:\TEMP\xhtml2pdf\git\testrender\data/source\adobe_asian_lenguage_pack.html
Traceback (most recent call last):
  File "E:/TEMP/xhtml2pdf/git/testrender/testrender.py", line 294, in <module>
    main()
  File "E:/TEMP/xhtml2pdf/git/testrender/testrender.py", line 225, in main
    pdf, pages, diff = render_file(filename, output_dir, ref_dir, options)
  File "E:/TEMP/xhtml2pdf/git/testrender/testrender.py", line 108, in render_file
    create_diff_image(page['png'], page['ref'],
  File "E:/TEMP/xhtml2pdf/git/testrender/testrender.py", line 64, in create_diff_image
    diff_value = int(float(result.strip()))
ValueError: could not convert string to float: b"0compare.exe: profile 'icc': 'RGB ': RGB color space not permitted on grayscale PNG `E:\\TEMP\\xhtml2pdf\\git\\testrender\\output\\adobe_asian_lenguage_pack.page0.diff.png' @ warning/png.c/MagickPNGWarningHandler/1748."

Seems like imagemagick is throwing a warning back to testrender.py instead of the result. And testrender.py  then tries to convert the warning into a float, which of course fails.

Do you know how to silence imagemagick and suppress the warnings? Or probably it would be even better to set up testrender.py correctly, so that it gives imagemagick the correct color space for the grayscale PNG.

Do you get the same warning/error? Maybe it also depends on the version. I'm using (more or less) the newest one: ImageMagick-7.0.10-31-portable-Q16-x64

Reply:
to avoid these warning alerts you have to add ``-quiet`` as the following image. 
![image](https://user-images.githubusercontent.com/69447117/95912532-298c1580-0d60-11eb-8bd8-34c5af7f2502.png)
 yes, I agree with you, could be the version. When I used testrender on xenial Image that use other version I don't have to add ``-quiet``, so when you finish your local test on windows you have to remove it.

Reply:
@pedroszg

Thanks, adding the `-quiet` indeed works. I did some more research and it seems like the better way would be to add
`-colorspace RGB` to the compare command. See: https://unix.stackexchange.com/questions/485292/rgb-color-space-not-permitted-on-grayscale Then the warning won't be showing at all.

`    _,result = exec_cmd(options, options.compare_cmd, '-metric', 'ae', srcfile1, srcfile2,'-lowlight-color','white',  '-colorspace', 'RGB', outfile)`

Would that as well be an option? Could we change that in the testrender.py file? Let me know what you think. 😊

Reply:
Also I'm struggling a bit with running the testrender.py script. If I'm running it from the "main" directory
`python testrender/testrender.py`, I get the following error:

Traceback (most recent call last):
  File "testrender\testrender.py", line 13, in <module>
    from xhtml2pdf import pisa
ModuleNotFoundError: No module named 'xhtml2pdf' 

I manually have to add:
 `sys.path.insert(0, os.getcwd())` to testrender.py to get it running. So it will add the main folder to the Python Path to find the xhtml2pdf module.

Should I make a new commit and maybe add that line?